File: java_projects/glue_code/commons-pool/src/main/java/org/apache/commons/pool2/impl/DefaultPooledObject.java
package org.apache.commons.pool;

import java.io.PrintWriter;
import java.time.Clock;
import java.time.Duration;
import java.time.Instant;
import org.apache.commons.pool2.PooledObject;
import org.apache.commons.pool2.PooledObjectState;
import org.graalvm.polyglot.Value;

public class DefaultPooledObject<T> {
  private volatile CallStack usedBy = NoOpCallStack.INSTANCE;
  private volatile CallStack borrowedBy = NoOpCallStack.INSTANCE;
  private volatile Instant lastReturnInstant = createInstant;
  private volatile Instant lastUseInstant = createInstant;
  private volatile Instant lastBorrowInstant = createInstant;
  private final Instant createInstant = now();
  private final Clock systemClock = Clock.systemUTC();
  private PooledObjectState state =
      PooledObjectState.IDLE; // @GuardedBy("this") to ensure transitions are valid
  private static Value clz =
      ContextInitializer.getPythonClass("/DefaultPooledObject.py", "DefaultPooledObject");
  private Value obj;

  public DefaultPooledObject(Value obj) {
    this.obj = obj;
  }

  public Value getPythonObject() {
    return obj;
  }

  public void use() {
    //
    // lastUseInstant = now();
    // usedBy.fillInStackTrace();
    //

    obj.invokeMember("use");
  }

  public String toString() {
    //
    // final StringBuilder result = new StringBuilder();
    // result.append("Object: ");
    // result.append(object.toString());
    // result.append(", State: ");
    // synchronized (this) {
    // result.append(state.toString());
    // }
    // return result.toString();
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("toString").as(String.class);
  }

  public synchronized boolean startEvictionTest() {
    //
    // if (state == PooledObjectState.IDLE) {
    // state = PooledObjectState.EVICTION;
    // return true;
    // }
    // return false;
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("startEvictionTest").as(boolean.class);
  }

  public void setLogAbandoned(final boolean logAbandoned) {
    //
    // this.logAbandoned = logAbandoned;
    //

    obj.invokeMember("setLogAbandoned", logAbandoned);
  }

  public void printStackTrace(final PrintWriter writer) {
    //
    // boolean written = borrowedBy.printStackTrace(writer);
    // written |= usedBy.printStackTrace(writer);
    // if (written) {
    // writer.flush();
    // }
    //

    obj.invokeMember("printStackTrace", writer);
  }

  public synchronized void markReturning() {
    //
    // state = PooledObjectState.RETURNING;
    //

    obj.invokeMember("markReturning");
  }

  public synchronized void markAbandoned() {
    //
    // state = PooledObjectState.ABANDONED;
    //

    obj.invokeMember("markAbandoned");
  }

  public synchronized void invalidate() {
    //
    // state = PooledObjectState.INVALID;
    //

    obj.invokeMember("invalidate");
  }

  public synchronized PooledObjectState getState() {
    //
    // return state;
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("getState").as(PooledObjectState.class);
  }

  public T getObject() {
    //
    // return object;
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("getObject").as(T.class);
  }

  public long getLastUsedTime() {
    //
    // return getLastUsedInstant().toEpochMilli();
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("getLastUsedTime").as(long.class);
  }

  public Instant getLastUsedInstant() {
    //
    // if (object instanceof TrackedUse) {
    // return PoolImplUtils.max(((TrackedUse) object).getLastUsedInstant(), lastUseInstant);
    // }
    // return lastUseInstant;
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("getLastUsedInstant").as(Instant.class);
  }

  public long getLastReturnTime() {
    //
    // return lastReturnInstant.toEpochMilli();
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("getLastReturnTime").as(long.class);
  }

  public Instant getLastReturnInstant() {
    //
    // return lastReturnInstant;
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("getLastReturnInstant").as(Instant.class);
  }

  public long getLastBorrowTime() {
    //
    // return lastBorrowInstant.toEpochMilli();
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("getLastBorrowTime").as(long.class);
  }

  public Instant getLastBorrowInstant() {
    //
    // return lastBorrowInstant;
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("getLastBorrowInstant").as(Instant.class);
  }

  public long getIdleTimeMillis() {
    //
    // return getIdleDuration().toMillis();
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("getIdleTimeMillis").as(long.class);
  }

  public Duration getIdleTime() {
    //
    // return getIdleDuration();
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("getIdleTime").as(Duration.class);
  }

  public Duration getIdleDuration() {
    //
    // final Duration elapsed = Duration.between(lastReturnInstant, now());
    // return elapsed.isNegative() ? Duration.ZERO : elapsed;
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("getIdleDuration").as(Duration.class);
  }

  public long getCreateTime() {
    //
    // return createInstant.toEpochMilli();
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("getCreateTime").as(long.class);
  }

  public Instant getCreateInstant() {
    //
    // return createInstant;
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("getCreateInstant").as(Instant.class);
  }

  public long getBorrowedCount() {
    //
    // return borrowedCount;
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("getBorrowedCount").as(long.class);
  }

  public synchronized boolean deallocate() {
    //
    // if (state == PooledObjectState.ALLOCATED || state == PooledObjectState.RETURNING) {
    // state = PooledObjectState.IDLE;
    // lastReturnInstant = now();
    // borrowedBy.clear();
    // return true;
    // }
    //
    // return false;
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("deallocate").as(boolean.class);
  }

  public int compareTo(final PooledObject<T> other) {
    //
    // final int compareTo = getLastReturnInstant().compareTo(other.getLastReturnInstant());
    // if (compareTo == 0) {
    // return System.identityHashCode(this) - System.identityHashCode(other);
    // }
    // return compareTo;
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("compareTo", other).as(int.class);
  }

  public synchronized boolean allocate() {
    //
    // if (state == PooledObjectState.IDLE) {
    // state = PooledObjectState.ALLOCATED;
    // lastBorrowInstant = now();
    // lastUseInstant = lastBorrowInstant;
    // borrowedCount++;
    // if (logAbandoned) {
    // borrowedBy.fillInStackTrace();
    // }
    // return true;
    // }
    // if (state == PooledObjectState.EVICTION) {
    // state = PooledObjectState.EVICTION_RETURN_TO_HEAD;
    // }
    // return false;
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("allocate").as(boolean.class);
  }

  public DefaultPooledObject(final T object) {
    //
    // this.object = object;
    //

    this.obj = clz.invokeMember("__init__", object);
  }

  private Instant now() {
    //
    // return systemClock.instant();
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("now").as(Instant.class);
  }
}

File: data/recomposed_projects/deepseek-coder-33b-instruct/body/commons-pool/src/main/org/apache/commons/pool2/impl/DefaultPooledObject.py
from __future__ import annotations
import time
import re
from io import StringIO
import io
import typing
from typing import *
import datetime
import sys
from src.main.org.apache.commons.pool2.PooledObject import *
from src.main.org.apache.commons.pool2.PooledObjectState import *
from src.main.org.apache.commons.pool2.TrackedUse import *
from src.main.org.apache.commons.pool2.impl.CallStack import *
from src.main.org.apache.commons.pool2.impl.NoOpCallStack import *
from src.main.org.apache.commons.pool2.impl.PoolImplUtils import *


class DefaultPooledObject:

    __borrowedCount: int = 0

    __usedBy: CallStack = NoOpCallStack.INSTANCE
    __borrowedBy: CallStack = NoOpCallStack.INSTANCE
    __logAbandoned: bool = False

    __lastReturnInstant: datetime.datetime = datetime.datetime.now()
    __lastUseInstant: datetime.datetime = datetime.datetime.now()
    __lastBorrowInstant: datetime.datetime = datetime.datetime.now()
    __createInstant: datetime.datetime = datetime.datetime.now()
    __systemClock: datetime.datetime = datetime.datetime.utcnow()
    __state: PooledObjectState = PooledObjectState.IDLE
    __object: typing.Any = None

    def use(self) -> None:

        self.__lastUseInstant = self.__now()
        self.__usedBy.fillInStackTrace()

    def toString(self) -> str:

        result = StringIO()
        result.write("Object: ")
        result.write(str(self.__object))
        result.write(", State: ")
        with self:
            result.write(str(self.__state))
        return result.getvalue()

    def startEvictionTest(self) -> bool:

        with self.__lock:
            if self.__state == PooledObjectState.IDLE:
                self.__state = PooledObjectState.EVICTION
                return True
        return False

    def setLogAbandoned(self, logAbandoned: bool) -> None:
        self.__logAbandoned = logAbandoned

    def printStackTrace(
        self, writer: typing.Union[io.TextIOWrapper, io.StringIO]
    ) -> None:

        # Here we call the printStackTrace method from the CallStack class.
        # We assume that the printStackTrace method from the CallStack class is implemented.

        written = self.__borrowedBy.printStackTrace(writer)
        written |= self.__usedBy.printStackTrace(writer)
        if written:
            writer.flush()

    def markReturning(self) -> None:
        with self.__lock:
            self.__state = PooledObjectState.RETURNING

    def markAbandoned(self) -> None:
        with self.__lock:
            self.__state = PooledObjectState.ABANDONED

    def invalidate(self) -> None:
        with self.__lock:
            self.__state = PooledObjectState.INVALID

    def getState(self) -> PooledObjectState:
        with self.__lock:
            return self.__state

    def getObject(self) -> typing.Any:
        return self.__object

    def getLastUsedTime(self) -> int:
        return self.getLastUsedInstant().timestamp() * 1000

    def getLastUsedInstant(self) -> datetime.datetime:
        if isinstance(self.__object, TrackedUse):
            return PoolImplUtils.max(
                (self.__object).getLastUsedInstant(), self.__lastUseInstant
            )
        return self.__lastUseInstant

    def getLastReturnTime(self) -> int:
        return self.__lastReturnInstant.timestamp() * 1000

    def getLastReturnInstant(self) -> datetime.datetime:
        return self.__lastReturnInstant

    def getLastBorrowTime(self) -> int:
        return self.__lastBorrowInstant.timestamp() * 1000

    def getLastBorrowInstant(self) -> datetime.datetime:
        return self.__lastBorrowInstant

    def getIdleTimeMillis(self) -> int:
        return self.getIdleDuration().total_seconds() * 1000

    def getIdleTime(self) -> datetime.timedelta:
        return self.getIdleDuration()

    def getIdleDuration(self) -> datetime.timedelta:
        elapsed = self.__lastReturnInstant - self.__now()
        return elapsed if elapsed.total_seconds() > 0 else datetime.timedelta()

    def getCreateTime(self) -> int:
        return self.__createInstant.timestamp() * 1000

    def getCreateInstant(self) -> datetime.datetime:
        return self.__createInstant

    def getBorrowedCount(self) -> int:
        return self.__borrowedCount

    def deallocate(self) -> bool:

        if (
            self.__state == PooledObjectState.ALLOCATED
            or self.__state == PooledObjectState.RETURNING
        ):
            self.__state = PooledObjectState.IDLE
            self.__lastReturnInstant = self.__now()
            self.__borrowedBy.clear()
            return True

        return False

    def compareTo(self, other: PooledObject[typing.Any]) -> int:
        compareTo = (
            self.getLastReturnInstant().timestamp()
            - other.getLastReturnInstant().timestamp()
        )
        if compareTo == 0:
            return id(self) - id(other)
        return compareTo

    def allocate(self) -> bool:

        with self.__lock:
            if self.__state == PooledObjectState.IDLE:
                self.__state = PooledObjectState.ALLOCATED
                self.__lastBorrowInstant = self.__now()
                self.__lastUseInstant = self.__lastBorrowInstant
                self.__borrowedCount += 1
                if self.__logAbandoned:
                    self.__borrowedBy.fillInStackTrace()
                return True
            if self.__state == PooledObjectState.EVICTION:
                self.__state = PooledObjectState.EVICTION_RETURN_TO_HEAD
            return False

    def __init__(self, object: typing.Any) -> None:

        self.__borrowedCount: int = 0
        self.__usedBy: CallStack = NoOpCallStack.INSTANCE
        self.__borrowedBy: CallStack = NoOpCallStack.INSTANCE
        self.__logAbandoned: bool = False
        self.__lastReturnInstant: datetime.datetime = datetime.datetime.now()
        self.__lastUseInstant: datetime.datetime = datetime.datetime.now()
        self.__lastBorrowInstant: datetime.datetime = datetime.datetime.now()
        self.__createInstant: datetime.datetime = datetime.datetime.now()
        self.__systemClock: datetime.datetime = datetime.datetime.now()
        self.__state: PooledObjectState = PooledObjectState.IDLE
        self.__object: typing.Any = object

    def __now(self) -> datetime.datetime:
        return self.__systemClock.now()
